File: Problemas/Arreglos/Container-with-Most-Water/Container-with-Most-Water.cpp
#include "../../../Libraries/ArrayPrinter.h"
#include <iostream>
#include <vector>
#include <climits>

// Fuerza Bruta O(n²)
// class Solution {
// public:
//     int maxArea(std::vector<int> &A)
//     {
//         int N = A.size(), ans = 0;
//         for (int i = 0; i < N; ++i)
//         {
//             if (!A[i])
//                 continue;
//             for (int j = i + 1 + ans / A[i]; j < N; ++j)
//             {
//                 ans = std::max(ans, std::min(A[i], A[j]) * (j - i));
//             }
//         }
//         return ans;
//     }
// };

// Dos Punteros O(n)
class Solution {
public:
    int maxArea(std::vector<int> &A)
    {
        int ans = 0, L = 0, R = A.size() - 1;
        while (L < R)
        {
            ans = std::max(ans, (R - L) * std::min(A[L], A[R]));
            if (A[L] < A[R])
                ++L; // Mover el borde más pequeño
            else
                --R;
        }
        return ans;
    }
};

int main()
{
    std::vector<int> height = {1, 8, 6, 2, 5, 4, 8, 3, 7};

    Solution sol;

    printArray(height);
    std::cout << "Output: " << sol.maxArea(height) << std::endl;

    return 0;
}

File: Problemas/Arreglos/Container-with-Most-Water/Container-with-Most-Water.cs
using System;
using System.Collections.Generic;

namespace ContainerWithMostWater
{
    public class Solution 
    {
        public int MaxArea(int[] height)
        {
            int ans = 0, L = 0, R = height.Length - 1;
            while (L < R)
            {
                ans = Math.Max(ans, (R - L) * Math.Min(height[L], height[R]));
                if (height[L] < height[R]) L++; // Move the smaller edge
                else R--;
            }
            return ans;
        }
        public static void Main(String[] args)
        {
            int[] height = new int[] {1,8,6,2,5,4,8,3,7};

            Console.WriteLine("Input: height = [" + string.Join(" ", height) + "]"); 

            int ans = new Solution().MaxArea(height);
            Console.WriteLine("Output: " + ans);


        }
    }   
}
